nodejs/aws-infra/experimental/cluster.ts

```diff
@@ -0,0 +1,141 @@
+
+import { Network } from "./../network";
+
+import * as utils from "../utils";
+
+/**
+ * A Cluster is a general purpose ECS cluster configured to run in a provided Network.
+ */
+export class Cluster extends pulumi.ComponentResource {
+    public readonly instance: aws.ecs.Cluster;
+
+    /**
+     * The network in which to create this cluster.
+     */
+    public readonly network: Network;
+    /**
+     * Security groups associated with this this ECS Cluster.
+     */
+    public readonly instanceSecurityGroups: aws.ec2.SecurityGroup[];
+
+    constructor(name: string, args: ClusterArgs, opts: pulumi.ComponentResourceOptions = {}) {
+        super("aws-infra:x:Cluster", name, {
+            ...args,
+        }, opts);
+
+        // First create an ECS cluster.
+        const parentOpts = { parent: this };
+        const instance = args.instance || new aws.ecs.Cluster(name, args, parentOpts);
+
+        const network = args.network || Network.getDefault();
+
+        // IDEA: Can we re-use the network's default security group instead of creating a specific
+        // new security group in the Cluster layer?  This may allow us to share a single Security Group
+        // across both instance and Lambda compute.
+        const instanceSecurityGroups = args.instanceSecurityGroups || [new aws.ec2.SecurityGroup(name, {
+            vpcId: network.vpcId,
+            ingress: Cluster.defaultSecurityGroupIngress(),
+            egress: Cluster.defaultSecurityGroupEgress(),
+            tags: { Name: name },
+        }, parentOpts)];
+
+        this.instance = instance;
+        this.network = network;
+        this.instanceSecurityGroups = instanceSecurityGroups;
+
+        this.registerOutputs({
+            instance,
+            network,
+            instanceSecurityGroups,
+        });
+    }
+
+    public static defaultSecurityGroupEgress() {
+        return [{
+            fromPort: 0,
+            toPort: 0,
+            protocol: "-1",  // all
+            cidrBlocks: [ "0.0.0.0/0" ],
+        }];
+    }
+
+    public static defaultSecurityGroupIngress() {
+        return [
+            // Expose SSH
+            {
+                fromPort: 22,
+                toPort: 22,
+                protocol: "TCP",
+                cidrBlocks: [ "0.0.0.0/0" ],
+            },
+            // Expose ephemeral container ports to Internet.
+            // TODO: Limit to load balancer(s).
+            {
+                fromPort: 0,
+                toPort: 65535,
+                protocol: "TCP",
+                cidrBlocks: [ "0.0.0.0/0" ],
+            },
+        ];
+    }
+}
+
+(<any>Cluster).doNotCapture = true;
+
+// The shape we want for ClusterArgs.  We don't export this as 'Overwrite' types are not pleasant to
+// work with. However, they internally allow us to succinctly express the shape we're trying to
+// provide. Code later on will ensure these types are compatible.
+type OverwriteShape = utils.Overwrite<aws.ecs.ClusterArgs, {
```

Comment:
I didn't follow most of this ... Does this have the effect of adding all properties on `aws.ecs.ClusterArgs` to the `ClusterArgs` type below or something? I notice we pass `args` above to the `aws.ecs.Cluster` constructor which led me to wonder if that means we expose the full superset of all arguments available on the ECS args?

Reply:
First, this is an internal detail for bookkeeping purposes.  The idea here is to provide a way (internally) to simple express the shape of data we want to expose.  IN this example, we are saying "the args to the Pulumi-Cluster shuld be equivlent to the args for the ecs-Cluster, except with the following changes made".  It allows us to simply see how we're augmenting the shape of things.

However, the 'Overwrite' type is less pleasant to use from a 'docs'/'intellisense'/go-to-def/'errors/diagnostics' perspective.  So we don't expose that type for users to use themselves.  Instead, we hand-write the type they will be using.  We then ensure that these types are comptible by doing a no-op assignment each type to the other.  in this way, we will discover if changes happen to one that are not reflected in the other.

Reply:
Second:

> I notice we pass args above to the aws.ecs.Cluster constructor which led me to wonder if that means we expose the full superset of all arguments available on the ECS args?

Right.  the intent here is to ensure we do not lock users out from being able to pass any args properties along to the underlying ecs.Cluster that is instantiated.  This is an 'escape hatch' if you will, that means that if we miss something, or are behind, users can still be unblocked and can pass along the data they need.